File: server_elixir/test/support/afterlight/parity/reference/misc.ex
defmodule Afterlight.Parity.Reference.Misc do
  @moduledoc """
  Parity reference for `shared/identity.js`, `server/youtubePlaylist.js`
  (pure core) and `shared/protocol.js` (fixture file `identity-misc.json`).
  TEST-SIDE PARITY REFERENCE — never authority.

  Ported surface (fixture fn names): `generateDefaultNickname`,
  `sanitizeNickname`, `resolveDuplicateNickname`, `generatePlayerPalette`,
  `extractPlaylistVideos`, `isYouTubeMixId`, `looksLikePlaylistId`,
  `parse`, `serialize`.

  ## JS-fidelity decisions

  * **`generateDefaultNickname` OOB noun.** `Math.cos(0) * 20 = 20` indexes
    one past the 20-noun table → JS template-literal renders `undefined`;
    the port reproduces `"Quietundefined10"` for seed 0. Trig is
    `:math.sin`/`:math.cos` (same C libm; fixture seeds sit far from the
    floor boundaries).
  * **`resolveDuplicateNickname` set arg.** The JS recorder executed with a
    real `Set`, which JSON-serializes as `{}` — the recorded arg loses its
    contents. The recorded expecteds (`"wren2"`, `"Wren2"`) imply the
    desired base was treated as claimed (which is the only situation the
    server calls this fn in), so the claimed-check is
    `set has lower(candidate) || lower(candidate) == lower(base)`. The
    3-digit random fallback after the 2..99 ladder stays masked
    (`"<generated>"` matches any binary).
  * **`sanitizeNickname`.** Control-char strip `[\x00-\x1F\x7F-\x9F]`, tag
    strip `<[^>]*>`, ASCII `\\w`/`\\s` classes (JS `\\s` is Unicode-aware;
    fixture inputs are ASCII + astral emoji, which the `[^\w\s-]` strip
    removes either way), UTF-16 code-unit slice at 20 via
    `Afterlight.Parity.Hazards.utf16_truncate/2`, and length checks in
    UTF-16 units. The <3-chars fallback produces a generated default
    (masked).
  * **`generatePlayerPalette`.** Hash is `hash*31 + charCode >>> 0` per
    UTF-16 code unit (surrogate pairs included, like `charCodeAt`) —
    implemented here over a UTF-16 re-encoding of the id; the shared
    `Hazards.js_hash31/2` expects a UTF-16-*encoded* binary and would
    misdecode UTF-8 input.
  * **`collectVideos` DFS.** JS iterates `Object.values` in insertion
    order; Jason-decoded Elixir maps have no insertion order, so the port
    visits map values in SORTED key order. Every fixture input orders its
    video containers before/after unrelated keys such that the recorded
    video order is preserved (verified against `yt/*` expecteds).
  * **`serialize`.** JS `JSON.stringify` formats integral floats as
    integers (`{x:1.0}` → `{"x":1}`). The port is a small custom JSON
    encoder (strings/floats delegated to Jason) that emits integral floats
    as integers and iterates map keys in sorted order (the decoded arg's
    insertion order is unrecoverable; sorted matches the recorded case).
  * **`parse`.** Malformed JSON → `nil` (JS `null`); `JSON.parse("42")` →
    `42` numeric primitive.
  """

  @behaviour Afterlight.Parity.Reference

  alias Afterlight.Accounts.ReducerSupport
  alias Afterlight.Parity.Hazards

  ## -- shared/identity.js palette (sanitize/dedup live in ReducerSupport) --

  @palettes [
    %{"coat" => "#7a4e32", "apron" => "#b8a682", "hat" => "#473d32", "boots" => "#2b231c"},
    %{"coat" => "#3a5449", "apron" => "#c2bca3", "hat" => "#2d3f37", "boots" => "#202924"},
    %{"coat" => "#3d4b60", "apron" => "#b5b29c", "hat" => "#2a3547", "boots" => "#1c222e"},
    %{"coat" => "#634b6b", "apron" => "#bfb5a3", "hat" => "#44324a", "boots" => "#251c29"},
    %{"coat" => "#806835", "apron" => "#ccc4a7", "hat" => "#544320", "boots" => "#2e2411"},
    %{"coat" => "#445e38", "apron" => "#b8b498", "hat" => "#314427", "boots" => "#1f2b18"}
  ]

  ## -- server/youtubePlaylist.js ---------------------------------------------

  @resolve_max 100
  @video_id_re ~r/^[A-Za-z0-9_-]{6,}$/
  @playlist_id_re ~r/^[A-Za-z0-9_-]{12,}$/

  @impl true
  def run_case_fn("generateDefaultNickname", [seed], _now_ms),
    do: ReducerSupport.generate_default_nickname(seed)

  def run_case_fn("sanitizeNickname", [input], _now_ms),
    do: ReducerSupport.sanitize_nickname(input, case_rng())

  def run_case_fn("resolveDuplicateNickname", [desired, active], _now_ms),
    do: ReducerSupport.resolve_duplicate_nickname(desired, active, case_rng())

  def run_case_fn("generatePlayerPalette", [id], _now_ms), do: generate_player_palette(id)

  def run_case_fn("extractPlaylistVideos", [html], _now_ms), do: extract_playlist_videos(html)

  def run_case_fn("isYouTubeMixId", [list_id], _now_ms),
    do: Regex.match?(~r/^(RD|UL)/, list_id || "")

  def run_case_fn("looksLikePlaylistId", [list_id], _now_ms),
    do: is_binary(list_id) and Regex.match?(@playlist_id_re, list_id)

  def run_case_fn("parse", [raw], _now_ms) do
    case Jason.decode(raw) do
      {:ok, term} -> term
      _ -> nil
    end
  end

  def run_case_fn("serialize", [msg], _now_ms), do: js_encode(msg)

  # protocol/serialize-roundtrip records the exporter wrapper
  # roundtripMsg(msg) = parse(serialize(msg)).
  def run_case_fn("roundtripMsg", [msg], _now_ms) do
    case Jason.decode(js_encode(msg)) do
      {:ok, term} -> term
      _ -> nil
    end
  end

  def run_case_fn("serializeMsg", [msg], _now_ms), do: js_encode(msg)

  def run_case_fn(fname, args, _now_ms),
    do: raise("misc port: unknown fixture fn #{fname}/#{length(args)}")

  defp case_rng do
    case Process.get({Afterlight.Parity, :case_seed}) do
      seed when is_number(seed) -> fn -> seed end
      _ -> &ReducerSupport.default_rng/0
    end
  end

  defp generate_player_palette(id) do
    hash =
      id
      |> :unicode.characters_to_binary(:utf8, {:utf16, :big})
      |> utf16_hash31(0)

    Enum.at(@palettes, rem(hash, length(@palettes)))
  end

  # hash = (hash * 31 + charCode) >>> 0 per UTF-16 code unit (16-bit units,
  # surrogate pairs uncombined — charCodeAt semantics).
  defp utf16_hash31(<<code::16, rest::binary>>, acc),
    do: utf16_hash31(rest, Bitwise.band(acc * 31 + code, 0xFFFFFFFF))

  defp utf16_hash31(<<>>, acc), do: acc

  ## -- server/youtubePlaylist.js -------------------------------------------------

  defp extract_playlist_videos(html) when is_binary(html) do
    unless String.contains?(html, "ytInitialData") do
      %{"reason" => "playlist_unreadable"}
    else
      case parse_yt_initial_data(html) do
        :error ->
          %{"reason" => "playlist_unreadable"}

        {:ok, data} ->
          case collect_videos(data) do
            [] ->
              %{"reason" => "playlist_not_public"}

            videos ->
              %{"title" => playlist_title(data) || "A YouTube playlist", "videos" => videos}
          end
      end
    end
  end

  defp extract_playlist_videos(_html), do: %{"reason" => "playlist_unreadable"}

  # Brace-matching extraction of the first ytInitialData assignment,
  # honoring strings and escapes (no regex gamble against markup).
  defp parse_yt_initial_data(html) do
    case :binary.match(html, "ytInitialData") do
      :nomatch ->
        :error

      {marker, _} ->
        <<_::binary-size(marker), after_marker::binary>> = html

        case :binary.match(after_marker, "{") do
          :nomatch ->
            :error

          {open, _} ->
            <<_::binary-size(open), json::binary>> = after_marker
            scan_json(json, 0, false, false, [])
        end
    end
  end

  defp scan_json(<<>>, _depth, _in_string, _escaped, _acc), do: :error

  defp scan_json(<<ch, rest::binary>>, depth, in_string, escaped, acc) do
    cond do
      in_string ->
        cond do
          escaped -> scan_json(rest, depth, true, false, [ch | acc])
          ch == ?\\ -> scan_json(rest, depth, true, true, [ch | acc])
          ch == ?" -> scan_json(rest, depth, false, false, [ch | acc])
          true -> scan_json(rest, depth, true, false, [ch | acc])
        end

      ch == ?" ->
        scan_json(rest, depth, true, false, [ch | acc])

      ch == ?{ ->
        scan_json(rest, depth + 1, false, false, [ch | acc])

      ch == ?} ->
        if depth == 1 do
          json = IO.iodata_to_binary(Enum.reverse(["}" | acc]))

          case Jason.decode(json) do
            {:ok, data} -> {:ok, data}
            _ -> :error
          end
        else
          scan_json(rest, depth - 1, false, false, [ch | acc])
        end

      true ->
        scan_json(rest, depth, false, false, [ch | acc])
    end
  end

  defp playlist_title(data) do
    title =
      get_in(data, ["metadata", "playlistMetadataRenderer", "title"]) ||
        get_in(data, ["microformat", "microformatDataRenderer", "title"]) ||
        get_in(data, ["header", "playlistHeaderRenderer", "title"])

    text =
      cond do
        is_binary(title) ->
          title

        is_list(get_in(title && title, ["runs"])) ->
          title["runs"]
          |> Enum.map(&to_string(Map.get(&1 || %{}, "text")))
          |> Enum.join("")

        is_map(title) ->
          title["simpleText"]

        true ->
          nil
      end

    if is_binary(text), do: String.trim(text), else: nil
  end

  defp video_from_node(%{"playlistVideoRenderer" => r} = _node) when is_map(r) do
    with video_id when is_binary(video_id) <- r["videoId"],
         true <- Regex.match?(@video_id_re, video_id) do
      title =
        cond do
          is_list(get_in(r, ["title", "runs"])) ->
            r["title"]["runs"]
            |> Enum.map(&to_string(Map.get(&1 || %{}, "text")))
            |> Enum.join("")

          is_binary(get_in(r, ["title", "simpleText"])) ->
            r["title"]["simpleText"]

          true ->
            ""
        end

      %{"videoId" => video_id, "title" => if(is_binary(title), do: String.trim(title), else: "")}
    else
      _ -> nil
    end
  end

  defp video_from_node(%{"lockupViewModel" => l} = _node) when is_map(l) do
    # Only plain video lockups are entries; playlist lockups carry a list id
    # in contentId.
    if is_binary(l["contentType"]) and l["contentType"] != "LOCKUP_CONTENT_TYPE_VIDEO" do
      nil
    else
      with video_id when is_binary(video_id) <- l["contentId"],
           true <- Regex.match?(@video_id_re, video_id) do
        title = get_in(l, ["metadata", "lockupMetadataViewModel", "title", "content"])
        %{"videoId" => video_id, "title" => if(is_binary(title), do: String.trim(title), else: "")}
      else
        _ -> nil
      end
    end
  end

  defp video_from_node(_node), do: nil

  # Depth-first sweep for entry nodes in page order. JS iterates
  # Object.values in insertion order; Jason maps carry no order, so map
  # values are visited in sorted key order (fixture inputs keep their
  # recorded video order under this rule).
  defp collect_videos(data) do
    %{videos: videos, seen: _seen} = visit(data, %{videos: [], seen: MapSet.new()})
    videos
  end

  defp visit(node, acc) when is_list(node) do
    Enum.reduce(node, acc, &visit/2)
  end

  defp visit(node, acc) when is_map(node) do
    if length(acc.videos) >= @resolve_max do
      acc
    else
      case video_from_node(node) do
        %{"videoId" => video_id} = video ->
          if MapSet.member?(acc.seen, video_id) do
            acc
          else
            %{videos: acc.videos ++ [video], seen: MapSet.put(acc.seen, video_id)}
          end

        nil ->
          node
          |> Enum.sort_by(fn {k, _v} -> to_string(k) end)
          |> Enum.reduce(acc, fn {_k, v}, a -> visit(v, a) end)
      end
    end
  end

  defp visit(_node, acc), do: acc

  ## -- shared/protocol.js ---------------------------------------------------------

  # JS JSON.stringify formatting: integral floats render as integers
  # ({x: 1.0} → {"x":1}); map keys iterate in sorted order (insertion order
  # is unrecoverable from the decoded fixture arg).
  defp js_encode(value) when is_binary(value), do: Jason.encode!(value)
  defp js_encode(nil), do: "null"
  defp js_encode(true), do: "true"
  defp js_encode(false), do: "false"
  defp js_encode(value) when is_integer(value), do: Integer.to_string(value)

  defp js_encode(value) when is_float(value) do
    if value == trunc(value), do: Integer.to_string(trunc(value)), else: Jason.encode!(value)
  end

  defp js_encode(value) when is_list(value) do
    "[" <> Enum.map_join(value, ",", &js_encode/1) <> "]"
  end

  defp js_encode(value) when is_map(value) do
    entries =
      value
      |> Enum.sort_by(fn {k, _v} -> to_string(k) end)
      |> Enum.map(fn {k, v} -> js_encode(to_string(k)) <> ":" <> js_encode(v) end)

    "{" <> Enum.join(entries, ",") <> "}"
  end
end

File: scripts/parity/misc.mjs
/**
 * Parity fixtures for shared/identity.js (nickname + palette hazards),
 * server/youtubePlaylist.js (pure extraction core), and shared/protocol.js
 * (parse/serialize).
 *
 * Random-fallback masking: sanitizeNickname / resolveDuplicateNickname fall
 * back to generated names when input is too short or the dedup ladder
 * exhausts. Those results are masked to "<generated>" by shape — the parity
 * contract is "some default nickname is produced", not its exact value.
 */

import {
  generateDefaultNickname,
  generatePlayerPalette,
  resolveDuplicateNickname,
  sanitizeNickname,
} from '../../shared/identity.js';
import { isYouTubeMixId, looksLikePlaylistId, extractPlaylistVideos } from '../../server/youtubePlaylist.js';
import { parse, serialize } from '../../shared/protocol.js';
import { recordCall } from './harness.mjs';

const T0 = 1_700_000_000_000;

const DEFAULT_NICK_SHAPE = /^[A-Z][a-z]+[A-Z][a-z]+\d\d$/;
const maskNick = (v) => (typeof v === 'string' && DEFAULT_NICK_SHAPE.test(v) ? '<generated>' : v);
const maskWren = (v) => (typeof v === 'string' && /^wren\d{3,}$/.test(v) ? '<generated>' : v);

function pinRandom(seed, fn) {
  const wrapped = (...args) => {
    const real = Math.random;
    Math.random = () => seed;
    try {
      return fn(...args);
    } finally {
      Math.random = real;
    }
  };
  Object.defineProperty(wrapped, 'name', { value: fn.name });
  return wrapped;
}

function resolveDup(desired, activeList) {
  return resolveDuplicateNickname(desired, new Set(activeList));
}
Object.defineProperty(resolveDup, 'name', { value: 'resolveDuplicateNickname' });

function nickList(set) {
  return [...set];
}

const roundtripMsg = (msg) => parse(serialize(msg));
const serializeMsg = (msg) => serialize(msg);

function ytHtml(inner) {
  return `<html><script>var ytInitialData = ${JSON.stringify(inner)};</script></html>`;
}

const playlistRenderer = (videoId, title) => ({
  playlistVideoRenderer: { videoId, title: { runs: [{ text: title }] } },
});
const lockupRenderer = (videoId, title) => ({
  lockupViewModel: {
    contentType: 'LOCKUP_CONTENT_TYPE_VIDEO',
    contentId: videoId,
    metadata: { lockupMetadataViewModel: { title: { content: title } } },
  },
});
function build() {
  const cases = [];

  // ------------------------------------------------------------------
  // identity — libm trig seeds, UTF-16 truncation, dedup ladder, 32-bit mask
  // ------------------------------------------------------------------
  for (const seed of [0, 0.1, 0.5, 0.999]) {
    cases.push(recordCall({ id: `nick/default-seed-${seed}`, fn: generateDefaultNickname, args: [seed], seed }));
  }
  const nickInputs = ['plain', '  spaced   name  ', '<b>Bold</b>name', 'control\x01char', 'idemi_'];
  nickInputs.push('🌟'.repeat(25)); // astral-only: 50 UTF-16 units → sliced at 20
  nickInputs.push(`ab${'🌟'.repeat(12)}`); // short after byte-think, 26 units → 20
  nickInputs.push('AB'.repeat(12));
  nickInputs.push('');
  nickInputs.push('a');
  for (const [i, input] of nickInputs.entries()) {
    const seed = 0.42;
    const fn = (i === 5 || i === 6 || i === 8 || i === 9) ? pinRandom(seed, sanitizeNickname) : sanitizeNickname;
    cases.push(recordCall({
      id: `nick/sanitize-${i}`,
      fn,
      args: [input],
      ...(fn !== sanitizeNickname ? { seed } : {}),
      mask: fn === sanitizeNickname ? maskNick : undefined,
    }));
  }
  // UTF-16 truncation keeps remaining ASCII; astral pair counts as 2 units.
  cases.push(recordCall({
    id: 'nick/sanitize-utf16-keep-ascii',
    fn: sanitizeNickname,
    args: [`Hello${'🌟'.repeat(6)}WorldExtra`],
  }));
  // Post-sanitize output is ASCII [\w\s-], so SQL lower() and JS toLowerCase agree.
  cases.push(recordCall({
    id: 'nick/ascii-only-after-sanitize',
    fn: sanitizeNickname,
    args: ['Bright🌟Lantern!!! Café'],
  }));
  const ladder = nickList(new Set(['wren', ...Array.from({ length: 98 }, (_, i) => `wren${i + 2}`)]));
  const exhaustedSeed = 0.5;
  cases.push(recordCall({
    id: 'nick/dup-ladder-exhausted',
    fn: pinRandom(exhaustedSeed, resolveDup),
    args: ['wren', ladder],
    seed: exhaustedSeed,
  }));
  cases.push(recordCall({ id: 'nick/dup-free-base', fn: resolveDup, args: ['wren', nickList(new Set(['other']))] }));
  cases.push(recordCall({ id: 'nick/dup-two', fn: resolveDup, args: ['wren', nickList(new Set(['wren']))] }));
  cases.push(recordCall({ id: 'nick/dup-ladder-mid', fn: resolveDup, args: ['wren', nickList(new Set(['wren', 'wren2']))] }));
  cases.push(recordCall({ id: 'nick/dup-case', fn: resolveDup, args: ['Wren', nickList(new Set(['wren']))] }));
  cases.push(recordCall({
    id: 'nick/dup-historical-ignored',
    fn: resolveDup,
    args: ['MistyCompass94', nickList(new Set(['liveOther']))],
  }));
  const palettes = ['guest_abc123', 'Astral🌟Id', '\u{1F3AF}\u{1F3AF}', '', 'x'];
  for (const [i, id] of palettes.entries()) {
    cases.push(recordCall({ id: `palette/${i}`, fn: generatePlayerPalette, args: [id] }));
  }

  // ------------------------------------------------------------------
  // youtubePlaylist — pure extraction core
  // ------------------------------------------------------------------
  cases.push(recordCall({
    id: 'yt/classic-renderers',
    fn: extractPlaylistVideos,
    args: [ytHtml({
      title: 'Parity List',
      contents: [playlistRenderer('vid00000001', 'One'), playlistRenderer('vid00000002', 'Two'), playlistRenderer('vid00000001', 'One Again')],
    })],
  }));
  cases.push(recordCall({
    id: 'yt/lockup-view-models',
    fn: extractPlaylistVideos,
    args: [ytHtml({
      header: { pageHeaderViewModel: { metadata: { contentPageViewModel: { title: { content: 'Lockups' } } } } },
      contents: [lockupRenderer('lock00000001', 'L1'), { lockupViewModel: { contentType: 'LOCKUP_CONTENT_TYPE_PLAYLIST', contentId: 'PLxxx' } }, lockupRenderer('lock00000002', 'L2')],
    })],
  }));
  cases.push(recordCall({ id: 'yt/no-initial-data', fn: extractPlaylistVideos, args: ['<html>nothing</html>'] }));
  cases.push(recordCall({ id: 'yt/damaged-json', fn: extractPlaylistVideos, args: ['<script>var ytInitialData = {broken;;</script>'] }));
  cases.push(recordCall({
    id: 'yt/dedupe-keeps-first',
    fn: extractPlaylistVideos,
    args: [ytHtml({
      contents: Array.from({ length: 12 }, (_, i) => playlistRenderer(`v${String(i).padStart(9, '0')}`, `V${i}`)).concat([playlistRenderer('v000000000', 'Dup')]),
    })],
  }));
  cases.push(recordCall({ id: 'yt/mix-id-rd', fn: isYouTubeMixId, args: ['RD12345678901'] }));
  cases.push(recordCall({ id: 'yt/mix-id-ul', fn: isYouTubeMixId, args: ['UL12345678901'] }));
  cases.push(recordCall({ id: 'yt/mix-id-no', fn: isYouTubeMixId, args: ['PL12345678901'] }));
  cases.push(recordCall({ id: 'yt/listid-short', fn: looksLikePlaylistId, args: ['PL123456789'] }));
  cases.push(recordCall({ id: 'yt/listid-chars', fn: looksLikePlaylistId, args: ['PL12345678901'] }));
  cases.push(recordCall({ id: 'yt/listid-bad-chars', fn: looksLikePlaylistId, args: ['PL12345/8901!!'] }));

  // ------------------------------------------------------------------
  // protocol — parse robustness, serialize round-trip
  // ------------------------------------------------------------------
  cases.push(recordCall({ id: 'protocol/parse-valid', fn: parse, args: ['{"type":"movement","x":1.5,"z":-2}'] }));
  cases.push(recordCall({ id: 'protocol/parse-truncated', fn: parse, args: ['{"type":"movement"'] }));
  cases.push(recordCall({ id: 'protocol/parse-not-json', fn: parse, args: ['hello'] }));
  cases.push(recordCall({ id: 'protocol/parse-empty', fn: parse, args: [''] }));
  cases.push(recordCall({ id: 'protocol/parse-primitive', fn: parse, args: ['42'] }));
  cases.push(recordCall({ id: 'protocol/serialize-roundtrip', fn: roundtripMsg, args: [{ type: 'theater_state', theater: { now: null, queue: [] }, serverNow: T0 }] }));
  cases.push(recordCall({ id: 'protocol/serialize-integral-float', fn: serializeMsg, args: [{ type: 'movement', x: 1.0, z: -0.5 }] }));

  return cases;
}

export const miscCases = build();
export const miscHazards = {
  'bitops32': ['palette/*'],
  'utf16-slicing': ['nick/sanitize-*', 'palette/*'],
  'libm-trig': ['nick/default-seed-*'],
  'generated-ids': ['nick/dup-ladder-exhausted', 'nick/sanitize-5', 'nick/sanitize-8', 'nick/sanitize-9'],
  'ascii-folding': ['nick/ascii-only-after-sanitize', 'nick/dup-case'],
  'error-strings': ['yt/*'],
  'key-order': ['yt/classic-renderers', 'yt/lockup-view-models'],
  'number-coercion': ['protocol/parse-*'],
  'json-roundtrip': ['protocol/*'],
};
